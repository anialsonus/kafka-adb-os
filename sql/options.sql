-- Test the extension validates OPTIONs properly
-- These tests do not require a Kafka instance. They can be run as common tests.

-- start_ignore
CREATE EXTENSION IF NOT EXISTS gp_inject_fault;

DROP EXTENSION IF EXISTS kadb_fdw CASCADE;
CREATE EXTENSION kadb_fdw;

SELECT gp_inject_fault('all', 'reset', dbid)
FROM gp_segment_configuration;
-- end_ignore


-- Test: Single unknown OPTION in SERVER

CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server;
-- end_ignore


-- Test: Multiple unknown OPTIONs in SERVER

CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    unknown_option_2 'unknown_value_2',
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server;
-- end_ignore


-- Test: Unknown OPTIONs do not transit to TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Single unknown OPTION in TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    unknown_option 'unknown_value'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Multiple unknown OPTIONs in TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    unknown_option_2 'unknown_value_2',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    unknown_option 'unknown_value'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: SELECT issues WARNINGs about unknown OPTION both in SERVER and in TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option_server 'unknown_value_server'
);

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    unknown_option_table 'unknown_value_table',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);

SELECT gp_inject_fault_infinite('kadb_fdw_inject_tuples', 'skip', dbid)
FROM gp_segment_configuration;

SELECT gp_inject_fault_infinite('kadb_fdw_partition_none', 'skip', dbid)
FROM gp_segment_configuration;
-- end_ignore

SELECT i, t FROM test_kadb_fdw_table;

-- start_ignore
SELECT gp_inject_fault('all', 'reset', dbid)
FROM gp_segment_configuration;

DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Invalid OPTION in SERVER

CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    k_automatic_offsets 'unknown_value'
);

-- start_ignore
DROP SERVER IF EXISTS test_kadb_fdw_server;
-- end_ignore


-- Test: Invalid OPTION in TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_automatic_offsets 'unknown_value',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Invalid boolean OPTION

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    format 'csv',
    csv_ignore_header 'unknown_value'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Empty int OPTION

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    format 'avro',
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '',
    k_timeout_ms '1000'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore


-- Test: Historical invalid OPTION in SERVER

CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    k_allow_offset_increase 'unknown_value'
);

-- start_ignore
DROP SERVER IF EXISTS test_kadb_fdw_server;
-- end_ignore


-- Test: Historical invalid OPTION in TABLE

-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore

CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_allow_offset_increase 'unknown_value',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);

-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore
