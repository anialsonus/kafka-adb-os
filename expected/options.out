-- Test the extension validates OPTIONs properly
-- These tests do not require a Kafka instance. They can be run as common tests.
-- start_ignore
CREATE EXTENSION IF NOT EXISTS gp_inject_fault;
NOTICE:  extension "gp_inject_fault" already exists, skipping
DROP EXTENSION IF EXISTS kadb_fdw CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to server test_kadb_fdw_distribution_server
drop cascades to foreign table test_kadb_fdw_t1
CREATE EXTENSION kadb_fdw;
SELECT gp_inject_fault('all', 'reset', dbid)
FROM gp_segment_configuration;
 gp_inject_fault 
-----------------
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
(8 rows)

-- end_ignore
-- Test: Single unknown OPTION in SERVER
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option'
-- start_ignore
DROP SERVER test_kadb_fdw_server;
-- end_ignore
-- Test: Multiple unknown OPTIONs in SERVER
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    unknown_option_2 'unknown_value_2',
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_2'
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option'
-- start_ignore
DROP SERVER test_kadb_fdw_server;
-- end_ignore
-- Test: Unknown OPTIONs do not transit to TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option 'unknown_value'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option'
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
NOTICE:  drop cascades to foreign table test_kadb_fdw_table
-- end_ignore
-- Test: Single unknown OPTION in TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    unknown_option 'unknown_value'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option'
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
NOTICE:  drop cascades to foreign table test_kadb_fdw_table
-- end_ignore
-- Test: Multiple unknown OPTIONs in TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    unknown_option_2 'unknown_value_2',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    unknown_option 'unknown_value'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_2'
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option'
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
NOTICE:  drop cascades to foreign table test_kadb_fdw_table
-- end_ignore
-- Test: SELECT issues WARNINGs about unknown OPTION both in SERVER and in TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    unknown_option_server 'unknown_value_server'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_server'
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    unknown_option_table 'unknown_value_table',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_table'
SELECT gp_inject_fault_infinite('kadb_fdw_inject_tuples', 'skip', dbid)
FROM gp_segment_configuration;
 gp_inject_fault_infinite 
--------------------------
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
(8 rows)

SELECT gp_inject_fault_infinite('kadb_fdw_partition_none', 'skip', dbid)
FROM gp_segment_configuration;
 gp_inject_fault_infinite 
--------------------------
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
(8 rows)

-- end_ignore
SELECT i, t FROM test_kadb_fdw_table;
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_table'
WARNING:  Kafka-ADB: Unknown OPTION 'unknown_option_server'
WARNING:  Kafka-ADB: Found no partitions in topic 'test_topic'
 i | t 
---+---
(0 rows)

-- start_ignore
SELECT gp_inject_fault('all', 'reset', dbid)
FROM gp_segment_configuration;
 gp_inject_fault 
-----------------
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
 Success:
(8 rows)

DROP SERVER test_kadb_fdw_server CASCADE;
NOTICE:  drop cascades to foreign table test_kadb_fdw_table
-- end_ignore
-- Test: Invalid OPTION in SERVER
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    k_automatic_offsets 'unknown_value'
);
ERROR:  k_automatic_offsets requires a Boolean value
-- start_ignore
DROP SERVER IF EXISTS test_kadb_fdw_server;
NOTICE:  server "test_kadb_fdw_server" does not exist, skipping
-- end_ignore
-- Test: Invalid OPTION in TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_automatic_offsets 'unknown_value',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);
ERROR:  k_automatic_offsets requires a Boolean value
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore
-- Test: Invalid boolean OPTION
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40',
    format 'csv',
    csv_ignore_header 'unknown_value'
);
ERROR:  csv_ignore_header requires a Boolean value
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore
-- Test: Empty int OPTION
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    format 'avro',
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_seg_batch '',
    k_timeout_ms '1000'
);
ERROR:  Kafka-ADB: 'k_seg_batch' OPTION must be an integer
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore
-- Test: Historical invalid OPTION in SERVER
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro',
    k_allow_offset_increase 'unknown_value'
);
WARNING:  Kafka-ADB: 'k_allow_offset_increase' OPTION is deprecated; use 'k_automatic_offsets' instead
ERROR:  k_automatic_offsets requires a Boolean value
-- start_ignore
DROP SERVER IF EXISTS test_kadb_fdw_server;
NOTICE:  server "test_kadb_fdw_server" does not exist, skipping
-- end_ignore
-- Test: Historical invalid OPTION in TABLE
-- start_ignore
CREATE SERVER test_kadb_fdw_server
FOREIGN DATA WRAPPER kadb_fdw
OPTIONS (
    k_brokers '0.0.0.0:9092',
    format 'avro'
);
-- end_ignore
CREATE FOREIGN TABLE test_kadb_fdw_table(i INT, t TEXT)
SERVER test_kadb_fdw_server
OPTIONS (
    k_topic 'test_topic',
    k_consumer_group 'test_consumer_group',
    k_allow_offset_increase 'unknown_value',
    k_seg_batch '3',
    k_timeout_ms '1000',
    k_tuples_per_partition_on_inject '2',
    k_initial_offset '40'
);
WARNING:  Kafka-ADB: 'k_allow_offset_increase' OPTION is deprecated; use 'k_automatic_offsets' instead
ERROR:  k_automatic_offsets requires a Boolean value
-- start_ignore
DROP SERVER test_kadb_fdw_server CASCADE;
-- end_ignore
